// enzyme-to-json/build/utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.compact = exports.includeInChildrenCompatible = exports.omitFromPropsCompatible = exports.includeInChildrenMinimal = exports.omitFromPropsMinimal = exports.isEnzymeWrapper = exports.isCheerioWrapper = exports.isReactWrapper = exports.isShallowWrapper = undefined;

var _lodash = require('lodash.filter');

var _lodash2 = _interopRequireDefault(_lodash);

var _lodash3 = require('lodash.isnil');

var _lodash4 = _interopRequireDefault(_lodash3);

var _ShallowWrapper = require('enzyme/build/ShallowWrapper');

var _ShallowWrapper2 = _interopRequireDefault(_ShallowWrapper);

var _ReactWrapper = require('enzyme/build/ReactWrapper');

var _ReactWrapper2 = _interopRequireDefault(_ReactWrapper);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SHALLOW_WRAPPER_NAME = _ShallowWrapper2.default.name;
var REACT_WRAPPER_NAME = _ReactWrapper2.default.name;

var isShallowWrapper = exports.isShallowWrapper = function isShallowWrapper(wrapper) {
  return wrapper && wrapper.constructor && wrapper.constructor.name === SHALLOW_WRAPPER_NAME;
};
var isReactWrapper = exports.isReactWrapper = function isReactWrapper(wrapper) {
  return wrapper && wrapper.constructor && wrapper.constructor.name === REACT_WRAPPER_NAME;
};
var isCheerioWrapper = exports.isCheerioWrapper = function isCheerioWrapper(wrapper) {
  return wrapper && wrapper.cheerio;
};

var isEnzymeWrapper = exports.isEnzymeWrapper = function isEnzymeWrapper(wrapper) {
  return isShallowWrapper(wrapper) || isReactWrapper(wrapper) || isCheerioWrapper(wrapper);
};

var omitFromPropsMinimal = exports.omitFromPropsMinimal = function omitFromPropsMinimal(val, key) {
  return key === 'children' || val === undefined;
};
var includeInChildrenMinimal = exports.includeInChildrenMinimal = function includeInChildrenMinimal(item) {
  return !(0, _lodash4.default)(item) && item !== '';
};

// For mountToDeepJson and mountToShallowJson
// to return test objects which are compatible with react-test-renderer
var omitFromPropsCompatible = exports.omitFromPropsCompatible = function omitFromPropsCompatible(val, key) {
  return key === 'children';
};
var includeInChildrenCompatible = exports.includeInChildrenCompatible = function includeInChildrenCompatible(item) {
  return !(0, _lodash4.default)(item);
};

var compact = exports.compact = function compact(array) {
  return (0, _lodash2.default)(array, includeInChildrenMinimal);
};
